### DISCUSSION TITLE
Get chain details from metamask?

### DISCUSSION THREAD

**Discussion by ackvf** [wevm/wagmi#1211] [Q&A]
During wagmi initialization, it is necessary to provide the chain definitions. Wagmi only comes with a handful of chains and we need to provide custom definitions for other chains we want to support.

https://wagmi.sh/docs/providers/configuring-chains#chains
```
import { chain, configureChains } from 'wagmi'
import { publicProvider } from 'wagmi/providers/public'

const { chains } = configureChains(
 [chain.mainnet, chain.optimism],
 [publicProvider()],
)
```

Instead of providing our own definitions I would like to rely on user's definitions configured in Metamask.
How can I query the chain details found there? I know I can use [`wallet_addEthereumChain
`](https://docs.metamask.io/guide/rpc-api.html#wallet-addethereumchain) to add new chain, but is there a way to acquire the information?

![image](https://user-images.githubusercontent.com/2313018/199135116-12b3b544-104c-4a74-89a9-d2bb1756e076.png)

**Comment by tmm** [comment]
Not that I'm aware of. We'll be adding more chains soon. What chains are you interested in?

**Reply by ackvf** [reply]
We eventually want to support many chains, but at this moment, when we are in beta, we only use testnets: Fuji, Goerli and Mumbai. So outside of your supported ones it's just Fuji 43113 I believe.

**Reply by ackvf** [reply]
But even if you do support (all) the chains out there, it might still be useful to be able to get the user's own preferred RPC endpoint as specified in Metamask. I wish this was supported by Metamask.

**Comment by ackvf** [comment]
@tmm So anyway, what we do currently is that for every supported chain, we first check if it's provided by wagmi and if not, we use chain definitions as provided by chainlist.org https://chainid.network/chains.json with this little mapping function

```ts
declare const chainlist: ChainDetails[]
declare const SUPPORTED_CHAIN_IDS: number[]

// --

const { chains, provider, webSocketProvider } = configureChains(
  getSupportedChains(), [
  publicProvider(),
])

// --

function getSupportedChains(): Chain[] {
  return SUPPORTED_CHAIN_IDS.map((id) => {

    /* First look if wagmi comes with our supported chains. */
    const wagmiChainDetails = allChains.find(v => v.id === id)
    if (wagmiChainDetails) return wagmiChainDetails

    /* As a fallback use our own definitions. */
    const chainDetails = chainlist.find(v => v.chainId === id)
    if (!chainDetails) throw 'Chain details missing.'

    /* We need to conform to the to the wagmi Chain format. */

    const { chainId, title, name, shortName, nativeCurrency, rpc, explorers, ens } = chainDetails as ChainDetails

    const blockExplorers = explorers?.reduce((p, { name, url }) => {

      p![name] = {
        name,
        url,
      }

      p!['default'] ??= p![name]

      return p

    }, {} as Chain['blockExplorers'])

    return {
      id: chainId,
      name: title || name,
      network: shortName,
      nativeCurrency,
      rpcUrls: { default: rpc[0], public: rpc[0] },
      blockExplorers,
      ens: ens?.registry && { address: ens?.registry },
      // multicall // no value in ChainDetails type
      // testnet: true, // `chainDetails.network` could be used, but it's not always defined.
    }
  })
}

// --

interface ChainDetails {
  name: string
  title?: string
  chain: string
  network?: string
  rpc: string[]
  faucets: string[]
  nativeCurrency: {
    name: string
    symbol: string
    decimals: number
  }
  infoURL: string
  shortName: string
  chainId: number
  /** @deprecated use chainId */
  networkId: number
  ens?: {
    registry: `0x${string}`
  }
  explorers?: [
    {
      name: string
      url: string
      standard: string
    }
  ]
}
```

The issue is that this is not always true. We sometimes use very specific custom RPC endpoints for cases where we need multichain support (querying data from multiple chains at the same time without switching chains in metamask) and those RPC endpoints are unusable as the primary RPC endpoint when used by wagmi, so that would lead to us having to maintain two lists, one for wagmi, one for our internal calls and yadi yada. So if we instead could maintain just one list and get the proper RPC endpoint from user's own definition from Metamask, it would be awesome.

Anyway, I do have a solution to our issue, it's just not so pretty at this moment.
